Add support for patching, merging, replacing metadata

hyperrealist opened the pull request:
* patch_metadata uses a http patch request with application/json-patch+json content type
* merge_metadata can be triggered with a http patch request and application/merge-patch+json
* update_metadata uses a similar approach to merge_metadata, but constructs a json-patch on the client-side
* add jsonpatch and json-merge-patch to requirements

danielballan commented on tiled/client/metadata_update.py:
```diff
@@ -0,0 +1,60 @@
+# adapted from https://github.com/OpenDataServices/json-merge-patch
+
+
+class DELETE_KEY:
```
A pattern that we use elsewhere in the codebase could be applied here: https://github.com/bluesky/tiled/blob/3b07e5f5add4d5ad2c5096a3694bee63d1a7308a/tiled/utils.py#L398-L415

hyperrealist replied:
Ah.. I have seen `Sentinel` before but totally forgot of its existence! Will use that instead. But perhaps it still makes sense to import `DELETE_KEY` from `tiled.client.metadata_update`? So I'll create `DELETE_KEY` there using `Sentinel`

danielballan commented on tiled/server/router.py:
```diff
@@ -1352,6 +1354,76 @@ async def put_awkward_full(
+
+    metadata_modified = False
+
+    # Specs should be ordered from most specific/constrained to least.
```
Can this block be factored out into a utility function instead of duplicated?

hyperrealist replied:
Will do. I think that is a sensible approach.

danielballan commented on tiled/client/base.py:
```diff
@@ -192,6 +196,12 @@ def metadata(self):
         # persistent.
         return DictView(self._item["attributes"]["metadata"])
 
+    def metadata_copy(self):
```
While we consider whether or not to keep `DictView` I'd suggest we hold off on adding this just yet. In the mean time we can recommend `copy.deepcopy(self.metadata)` in docstrings.

If we decide to keep `DictView`, we can add this in a follow-up PR. But once added it's harder to remove it.

danielballan replied:
Thinking about this more, and speaking with @Kezzsim, I'm appreciating that there isn't really a way around this. I've come around on keeping this method.

danielballan reviewed (commented):
Looks good. I think the main thing that's missing is implementing the new adapters methods on the SQL-backed adapter in Tiled itself. https://github.com/bluesky/tiled/blob/3b07e5f5add4d5ad2c5096a3694bee63d1a7308a/tiled/catalog/adapter.py#L814

In a later PR, we could even look into pushing the patch representation all the way into SQL, but to start I think applying the path in Python and doing a full replace in SQL is the way to go.

Kezzsim reviewed (commented):
Looked at this earlier on a call with Dan, not seeing anything too out of the ordinary that he hasn't already touched on.

hyperrealist commented:
@danielballan, @Kezzsim this PR is ready for another review. Changes since your last review are minimal, though it took me an obscene amount of time to figure this out.

genematx commented on tiled/server/schemas.py:
```diff
@@ -444,4 +445,42 @@ def specs_uniqueness_validator(cls, v):
+    specs: Optional[Specs]
+
+    # Wait for fix https://github.com/pydantic/pydantic/issues/3957
+    # to do this with `unique_items` parameters to `pydantic.constr`.
```
Looks like they've fixed it already in ver 1.10.12 (we use 1.10.13). Would it be worthwhile just uncommenting line 86?

genematx replied:
or maybe in a separate PR?

hyperrealist replied:
I marked this one as resolved by mistake. This came from a part of code I copied from @danielballan. Not exactly sure what needs to be done here, but I agree maybe it should be a separate PR.

https://github.com/bluesky/tiled/blame/e056aa9694375d2b034d69a0143d40845b6ee2bb/tiled/server/schemas.py#L436-L437

danielballan commented on tiled/client/base.py:
```diff
@@ -370,6 +380,99 @@ def update_metadata(self, metadata=None, specs=None):
+                self._item["attributes"]["metadata"] = content["metadata"]
+            else:
+                # Metadata was accepted as it is by the server.
+                # It is updated locally with the version submitted buy the client.
```
```suggestion
                # It is updated locally with the version submitted by the client.
```

danielballan commented on tiled/server/router.py:
```diff
@@ -1374,6 +1356,57 @@ async def put_awkward_full(
+
+    if request.headers["content-type"] == "application/json-patch+json":
+        metadata = apply_json_patch(
+            entry.metadata(), body.patch if body.patch is not None else []
```
This is a nice pattern when an empty collection (`[]` in this case) or `None` should both lead to the same result.

```suggestion
            entry.metadata(), (body.patch or [])
```

danielballan commented on tiled/server/router.py:
```diff
@@ -1374,6 +1356,57 @@ async def put_awkward_full(
+        )
+    elif request.headers["content-type"] == "application/merge-patch+json":
+        metadata = apply_merge_patch(
+            entry.metadata(), body.patch if body.patch is not None else {}
```
Same as above

```suggestion
            entry.metadata(), (body.patch or {})
```

danielballan commented on tiled/client/base.py:
```diff
@@ -370,6 +380,99 @@ def update_metadata(self, metadata=None, specs=None):
+        metadata : dict, optional
+            User metadata. May be nested. Must contain only basic types
+            (e.g. numbers, strings, lists, dicts) that are JSON-serializable.
+            Hint:
```
Consider replacing this explanation with an example:

```
Examples
--------

Make a mutable copy of the metadata, for editing.

>>> md = x.metadata_copy()

Add, remove, or change contents, e.g.

>>> md['a'] = 1
>>> del md['b']

Update the copy on the server.

>>> x.update_metadata(metadata)
```

Easier to quickly parse than an explanation.

danielballan commented on tiled/client/base.py:
```diff
@@ -370,6 +380,99 @@ def update_metadata(self, metadata=None, specs=None):
+        specs : List[str], optional
+            List of names that are used to label that the data and/or metadata
+            conform to some named standard specification.
+
```
Consider adding:

```
See Also
--------
patch_metadata
replace_metadata
```

And likewise on the other methods, to mutually cross-reference

danielballan commented on tiled/client/base.py:
```diff
@@ -370,6 +380,99 @@ def update_metadata(self, metadata=None, specs=None):
+            Hint:
+                You can use `metadata_copy()` to retrieve a mutable copy of
+                the current metadata that can be passed here with modifications.
+            Note:
```
The numpydoc standard supports a `Note` section. I'd suggest moving this out to keep the `Parameters` section streamlined.

I don't quite understand what this first sentence means. Maybe in the Note, explain, "This compares the input to the current metadata and constructs a JSON patch," and add additional example(s) to the Example section to illustrate `DELETE_KEY` and  this statement about relative paths.

hyperrealist replied:
"relative paths" was a bad word choice on my part.  I will add examples to demonstrate what I actually mean.

genematx commented on tiled/client/base.py:
```diff
@@ -370,6 +380,99 @@ def update_metadata(self, metadata=None, specs=None):
+        EXPERIMENTAL: Replace metadata.
+
+        This is subject to change or removal without notice
+
```
Do we need to make any changes to `metadata_revisions`, here or elsewhere? Or is it unrelated?

hyperrealist replied:
Good point. When you call `update_metadata` it creates a revision before updating the metadata/specs. https://github.com/bluesky/tiled/blob/3b07e5f5add4d5ad2c5096a3694bee63d1a7308a/tiled/catalog/adapter.py#L836-L844

`metadata_revisions` always queries latest revisions from the server. So I think we should be fine there.

genematx commented on tiled/client/base.py:
```diff
@@ -370,6 +380,99 @@ def update_metadata(self, metadata=None, specs=None):
 
         This is subject to change or removal without notice
 
+        Parameters
```
Would it be useful to expand the main description in the docstring a little bit. I guess, it's fairly obvious what the function is doing just from it's name, so maybe just emphasising that the logic here is the same as in the usual `old_dict.update(new_dict)`, i.e. some values would be overwritten or added, but none of them would be removed, whereas `replace_metadata` can lead to some key-value pairs being dropped. (At least this is my understanding of how it works; am I right?)

The distinction between `update_` and `patch_` metadata is less obvious, I guess.

hyperrealist replied:
I agree. I'll slightly expand the main description and add more details where appropriate.

hyperrealist commented:
I pushed a couple of commits addressing suggestions from @danielballan and @genematx

danielballan commented on tiled/client/base.py:
```diff
@@ -364,11 +374,139 @@ def formats(self):
+            dumps=orjson.dumps,
+        ).patch
+
+        if not exec:
```
I like the addition of a way to inspect the patch before applying it. I'd like to explore other ways to implement that, because this violates the principle that we generally try to follow: an optional parameter should not alter the return type.

I can think of two ways to achieve this:

1. Pull this out into a convenience function, `build_patch`.
2. Make it so that `update_metadata` returns the patch unconditionally, and have `exec` (or perhaps the inverse, `dry_run`) control whether or not it is applied, not what is returned. This might cut against the "user friendly" purpose though, as the user looking at the return value might wonder, "What am I supposed to _do_ with this? Did it work?" Perhaps this points toward (1).

hyperrealist replied:
Thanks @danielballan, as is always this is another very helpful comment. I "knew" I was violating some design principle, but couldn't figure which one. I am leaning toward the convenience function approach for the same reason you mention in 2.

danielballan reviewed (commented):
One note in line on the latest changes.

Also:
- @skarakuzu recently added a [CHANGELOG](https://github.com/bluesky/tiled/blob/main/CHANGELOG.md). Can you add a line item for this?
- Can these methods be added to the [API reference](https://github.com/bluesky/tiled/blob/c2f4e606309cb937336181c1fc9212b597c8d6a4/docs/source/reference/python-client.md?plain=1#L139)? It looks like `update_metadata` was never added, and the new methods should be added too.

danielballan commented:
I’ve asked @Kezzsim to test drive this locally a bit before we merge. Merge away when you are satisfied, Kari!

danielballan commented:
Rebased and force-pushed to resolve merge conflicts with #695

danielballan commented:
Tests failed due to an incompatibility with pydantic 2.x, introduced by #695. I have pushed a commit with a fix.

Kezzsim commented:
This looks good, however I'm noticing a break from convention when it comes to how HTTP is implemented. Pulling up swagger docs on http://127.0.0.1:8000/docs reveals that `application/json` is the only mimetype in the fastAPI dropdown.
![jsonpatchproblem1](https://github.com/bluesky/tiled/assets/15878154/9850ac26-db8c-4d87-8f94-0361ee7ba3dd)
 Technically this is correct, as the body in @hyperrealist 's spec is a json object which _contains_ valid json patches, but the overall request body itself is not a json patch. 
 
 Example of current request body [dict]:
 ```json
 {
  "patch": [
    {
      "op": "replace",
      "path": "/test",
      "value": "testing"
    }
  ],
  "specs": null
}
 ```
 Proposed json patch implementation of body [list]:
 ([conforming to RFC6902](https://datatracker.ietf.org/doc/html/rfc6902))
 ```json-patch+json
 [
    {
      "op": "replace",
      "path": "metadata/test",
      "value": "testing"
    }
  ]
 ```
 Fixing this to have it be within spec would require one of two potential changes:
 1) Split the endpoint into `PATCH /api/v1/metadata/{path}` and `PATCH /api/v1/specs/{path}` which both take valid json-patch objects that alter their respective properties at the root path.
 2) (As seen in the example above) change the body to be a valid json-patch but make the RFC6902 `path` roots be `metadata` and `specs` respectively.
 
 @danielballan mentioned pulling in @dylanmcreynolds since he has a good eye for making similar API design decisions that are both conforming to spec and aesthetically pleasing. I'd like to see what his suggestion is.

hyperrealist commented:
Thanks @Kezzsim, this is something that had been bothering me in the background at times. I appreciate your attention to detail!

>  1. Split the endpoint into `PATCH /api/v1/metadata/{path}` and `PATCH /api/v1/specs/{path}` which both take valid json-patch objects that alter their respective properties at the root path.
> 
>  2. (As seen in the example above) change the body to be a valid json-patch but make the RFC6902 `path` roots be `metadata` and `specs` respectively.

1. If it is typical to have to update both `metadata` and `specs` together this would necessitate extra http requests. If that is not the case I think this is a very much viable solution.
2. This is a creative way to conform to the http spec but I don't like how arguments are mixed together, especially since _how_ they are mixed together would change depending on the mimetype (we support both `application/json-patch+json` and `application/merge-patch+json`. See [here](https://github.com/bluesky/tiled/blob/2faa460a44da2b08fbc0f37592ec3815fbfa3812/tiled/server/router.py#L1397-L1405)).
3. A third solution might be to let the payload be of `application/json` mimetype and split the endpoint into something like `PATCH /api/v1/patch_metadata/{path}` and `PATCH /api/v1/merge_metadata/{path}`, each requiring a valid json with an embedded json patch or a merge patch respectively. But I think this is achieving little at the expense of polluting the API and adding to the complexity.
4. A fourth solution is to enforce a sub-mimetype at the content level:
```py
{
 "type": "application/json-patch+json",  # could default to this. may also support aliases  "json-patch"
 "metadata": [
   {
     "op": "replace",
     "path": "/test",
     "value": "testing"
   }
 ],
 "specs": [],
}
```
and
```py
{
 "type": "application/merge-patch+json",  # or "merge-patch" / "merge"
 "metadata": {"test": "testing"},
 "specs": [],
}
```

All these solutions would conform to http spec, but I am leaning toward solutions 1 or 4 for their conceptual simplicity and ease of implementation. Anyway, I am completely open to other ideas and design considerations.

danielballan commented:
Yes, good catch @Kezzsim! And thanks for presenting some additional options @hyperrealist.

Some additional design considerations:

Option (1) would break symmetry between GET and PATCH, as `GET /metadata/{path}` currently returns all "metadata" about a node, broadly defined, including structure family, structure, specs, data sources, and metadata. Making separate routes for PATCH seems asymmetrical in this sense, in addition to introducing overhead from separate requests for some use cases.

I can foresee adding support for patching data sources, alongside metadata and specs, which would make the downsides with (2) and (3) even more salient.

I can see no problems with (4). There was something appealing about specifying the patch format in the `content-type` header, but there is also something clean about making the request plain JSON (everybody knows JSON!) with the patch format specified in the body. It seems to keep our options open.

If we go that way, it is worth considering options for what to name the key. Is it `content-type`, `format`, `minetype`, `type`?

hyperrealist commented:
I very much intentionally wanted to avoid calling it `content-type`. I wouldn't envy whoever is documenting `content-type` the http header and `content-type` the subtype for this endpoint. `mimetype` sounds too restrictive to me as I would hesitate to support aliases under that name. Between `format` and `type` I like `type` because it hints at this being a mimetype without being overly restrictive. If adding a bit to the boilerplate is not an issue I would go with `patch-type`, `metadata-patch`, and `specs-patch`.

danielballan commented:
Seeing :+1:s from @Kezzsim and myself, I think going ahead with (4) with name `type` makes sense.

dylanmcreynolds commented:
> I very much intentionally wanted to avoid calling it content-type. I wouldn't envy whoever is documenting content-type the http header and content-type the subtype for this endpoint. mimetype sounds too restrictive to me as I would hesitate to support aliases under that name.

I'm not sure I understand the hesitance about calling it `content-type`. There's precedence here. For example, it normally appears in both the header and the body of multi-part form posts: https://swagger.io/docs/specification/describing-request-body/multipart-requests/

hyperrealist commented:
CI errors for Python 3.11 are coming from dask's incompatibility with 3.11.9 ([fixed in 2024.4.1](https://github.com/dask/dask/issues/11038)). Github CI uses 3.11.9, so it is broken at the moment.

I pinned python<3.11.9 for now, but I am not sure if I should attempt to resolve this issue in this PR. We should probably update to dask 2024.4.1.

danielballan commented on tiled/server/router.py:
```diff
@@ -1390,6 +1372,62 @@ async def put_awkward_full(
+            detail="This node does not support update of metadata.",
+        )
+
+    if body.content_type in [
```
In the spirit of, "Adhere to standards wherever possible," I would prefer to start with strictly accepting `application/json-patch+json` and `application/merge-patch+json` here. The use case for aliases seems narrow to me. The MIME type are unambiguous, Google-able, and easy enough to copy/paste from examples, or achieve through tooling like the Tiled Python client.

At least, I would wait to add aliases until/unless there is demonstrated demand for it that can be easily addressed another way. It's easy to add this later if we need to, but hard to remove once supported.

danielballan commented on tiled/server/schemas.py:
```diff
@@ -499,3 +500,45 @@ def specs_uniqueness_validator(cls, v):
+
+class HyphenizedBaseModel(pydantic.BaseModel):
+    # This model configuration allows aliases like "content-type"
+    model_config = ConfigDict(alias_generator=lambda f: f.replace("_", "-"))
```
Clever!

danielballan commented:
FYI, @hyperrealist, you may have seen this go by already but in case not, the CI failures on Python 3.11 are not the fault of this PR and can be ignored. See #715.

danielballan commented:
Rebased to pull in latest `CHANGELOG` and place this change entry on "Next" release.

danielballan commented:
Revisiting @hyperrealist's [comment](https://github.com/bluesky/tiled/pull/688#issuecomment-2022135250) proposing this API:

```py
{
 "type": "application/json-patch+json",  # could default to this. may also support aliases  "json-patch"
 "metadata": [
   {
     "op": "replace",
     "path": "/test",
     "value": "testing"
   }
 ],
 "specs": [],
}
```
and
```py
{
 "type": "application/merge-patch+json",  # or "merge-patch" / "merge"
 "metadata": {"test": "testing"},
 "specs": [],
}
```

I now realize that this was a little ambiguous about what's happening with `specs`. As currently implemented in this PR, `specs` is handled the same as in `PUT`. The content of `specs` in the request _replaces_ the current `specs`.

```py
class PatchMetadataRequest(HyphenizedBaseModel):
    content_type: str

    # These fields are optional because None means "no changes; do not update".
    patch: Optional[Union[List[JSONPatchSpec], Dict]]  # Dict for merge-patch
    specs: Optional[Specs]
```

I think it would be more useful (and more consistent) to make `specs` _also_ a patch.

```py
class PatchMetadataRequest(HyphenizedBaseModel):
    content_type: str

    # These fields are optional because None means "no changes; do not update".
    metadata: Optional[Union[List[JSONPatchSpec], Dict]]  # Dict for merge-patch
    specs: Optional[Union[List[JSONPatchSpec], Dict]]  # Dict for merge-patch
```

Notes:
- The user has to pick one patch `content-type`  for the whole request; they cannot be mixed. This seems like a good simplifying constraint.
- For specs, the performance advantage of "patch" vs "replace" is low. The list of `specs` is not expected to become long.
- But **in a future PR** I foresee adding `data_sources` here, and at that point I think we'll be glad for a consistent way to amend the `metadata`, `specs` , and `data_sources`, each with their own patch.

Existing unit tests cover `update_metadata`, but I think it's worthwhile to add a couple tests covering the new methods.
